### PR TITLE
added select-by capability based on mz and intensity

### PR DESCRIPTION
**Pull request by jspaaks**
- added tests
- updated module exports

refs #54

### CODE REVIEW

File: matchms/filtering/select_by_intensity.py [hunk 1]
```diff
@@ -0,0 +1,11 @@
+import numpy
+
+
+def select_by_intensity(spectrum, intensity_from=10.0, intensity_to=200.0):
```
Comment (fdiblen) [inline matchms/filtering/select_by_intensity.py:4]: can we add docstrings please?

File: matchms/filtering/select_by_mz.py [hunk 2]
```diff
@@ -0,0 +1,11 @@
+import numpy
+
+
+def select_by_mz(spectrum, mz_from=0.0, mz_to=1000.0):
```
Comment (fdiblen) [inline matchms/filtering/select_by_mz.py:4]: can we add docstrings please?

File: matchms/filtering/select_by_intensity.py (hunk 1 above)
Comment (jspaaks) [reply matchms/filtering/select_by_intensity.py:4]: can we do it later

File: matchms/filtering/select_by_mz.py (hunk 2 above)
Comment (jspaaks) [reply matchms/filtering/select_by_mz.py:4]: later

File: tests/test_select_by_intensity.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
+from matchms import Spectrum
+from matchms.filtering import select_by_intensity
+import numpy
+
+
+def test_select_by_intensity_no_bounds():
```
Comment (fdiblen) [inline tests/test_select_by_intensity.py]: can we add docstrings please?

File: tests/test_select_by_intensity.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+    mz = numpy.array([10, 20, 30, 40], dtype="float")
+    intensities = numpy.array([1, 10, 100, 1000], dtype="float")
+    spectrum = Spectrum(mz=mz, intensities=intensities, metadata=dict())
+
+    select_by_intensity(spectrum)
+
+    assert len(spectrum.mz) == 2
```
Comment (fdiblen) [inline tests/test_select_by_intensity.py]: numpy has shape and size functions. Please use these functions if you need to check the 'length'

File: tests/test_select_by_intensity.py [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+
+def test_select_by_intensity_lower_bound_only():
+
+    mz = numpy.array([10, 20, 30, 40], dtype="float")
+    intensities = numpy.array([1, 10, 100, 1000], dtype="float")
+    spectrum = Spectrum(mz=mz, intensities=intensities, metadata=dict())
+
+    select_by_intensity(spectrum, intensity_from=15.0)
+
+    assert len(spectrum.mz) == 1
```
Comment (fdiblen) [inline tests/test_select_by_intensity.py]: can you add assertion messages? something like in matchms/filtering/select_by_intensity.py

File: tests/test_select_by_intensity.py [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+
+    select_by_intensity(spectrum)
+
+    assert len(spectrum.mz) == 2
+    assert len(spectrum.mz) == len(spectrum.intensities)
+    assert numpy.array_equal(spectrum.mz, numpy.array([20, 30], dtype="float"))
+    assert numpy.array_equal(spectrum.intensities, numpy.array([10, 100], dtype="float"))
+
+
+def test_select_by_intensity_lower_bound_only():
```
Comment (fdiblen) [inline tests/test_select_by_intensity.py]: add docstring

File: tests/test_select_by_intensity.py [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+
+    select_by_intensity(spectrum, intensity_from=15.0)
+
+    assert len(spectrum.mz) == 1
+    assert len(spectrum.mz) == len(spectrum.intensities)
+    assert numpy.array_equal(spectrum.mz, numpy.array([30], dtype="float"))
+    assert numpy.array_equal(spectrum.intensities, numpy.array([100], dtype="float"))
+
+
+def test_select_by_intensity_upper_bound_only():
```
Comment (fdiblen) [inline tests/test_select_by_intensity.py:34]: add intensity_from to function arguments

File: tests/test_select_by_intensity.py (hunk 5 above)
Comment (jspaaks) [reply tests/test_select_by_intensity.py]: I will

File: tests/test_select_by_intensity.py [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+
+    select_by_intensity(spectrum, intensity_from=15.0, intensity_to=135.0)
+
+    assert len(spectrum.mz) == 1
+    assert len(spectrum.mz) == len(spectrum.intensities)
+    assert numpy.array_equal(spectrum.mz, numpy.array([30], dtype="float"))
+    assert numpy.array_equal(spectrum.intensities, numpy.array([100], dtype="float"))
+
+
+def test_select_by_intensity_no_bounds_2():
```
Comment (fdiblen) [inline tests/test_select_by_intensity.py]: more descriptive function name is needed

File: tests/test_select_by_intensity.py (hunk 4 above)
Comment (jspaaks) [reply tests/test_select_by_intensity.py]: I think we can agree this is a detail, right? The code is not wrong as it is, i.e. we do not have a problem in this actual moment. But, importantly, while I'm working to implement the requested changes, the rest of the team is waiting until they are starting to see the outline of the refactored structure.

When reviewing pull requests, I go by the following rules of thumb
1. Does the PR break any of the existing stuff
1. Is the PR an improvement over what is there already

if 1 is no and 2 is yes, we can merge.

File: tests/test_select_by_mz.py [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+
+    select_by_mz(spectrum, mz_from=15.0, mz_to=35.0)
+
+    assert len(spectrum.mz) == 2
+    assert len(spectrum.mz) == len(spectrum.intensities)
+    assert numpy.array_equal(spectrum.mz, numpy.array([20, 30], dtype="float"))
+    assert numpy.array_equal(spectrum.intensities, numpy.array([10, 100], dtype="float"))
+
+
+def test_select_by_mz_no_bounds_2():
```
Comment (fdiblen) [inline tests/test_select_by_mz.py]: find a better name

**Review by fdiblen — CHANGES_REQUESTED** [review]
@jspaaks  Nice!
Requested some changes. If you could add docstrings too all functions it would be great.
Do we want to have _select top X by intensity/mz_ function as well?
There is also a conflict in matchms/filtering/__init__.py

File: tests/test_select_by_intensity.py (hunk 5 above)
Comment (jspaaks) [reply tests/test_select_by_intensity.py]: I started work on adding assert messages but noticed that I was making it worse. Compare the two snippets below; the first one without assert messages, the second with assert messages added. Which do you find more readable?

```python
assert len(spectrum.mz) == 2, "Expected the spectrum's mz to have been filtered down to just 2 elements."
assert len(spectrum.mz) == len(spectrum.intensities), "Expected the spectrum's mz to be equal in length to its " \
                                                  "intensities."
assert numpy.array_equal(spectrum.mz, numpy.array([20, 30], dtype="float")), "Expected spectrum's mz array to be " \
                                                                         "equal to [20, 30]"
assert numpy.array_equal(spectrum.intensities, numpy.array([10, 100], dtype="float")), "Expected spectrum's "\
                                                                                   "intensities to be equal " \
                                                                                   "to [10, 100]"
```


```python
assert len(spectrum.mz) == 2
assert len(spectrum.mz) == len(spectrum.intensities)
assert numpy.array_equal(spectrum.mz, numpy.array([20, 30], dtype="float"))
assert numpy.array_equal(spectrum.intensities, numpy.array([10, 100], dtype="float"))
```

If readability is the only argument, I think we should omit assert messages if the message does not add to the statement. But perhaps there are other arguments that I'm forgetting?

**Comment by jspaaks** [comment]
- I'm reluctant to start adding docstrings to test functions; the ``filtering`` functions yes, we should do that (although not now, I'd like us to focus on laying out the general structure before filling in the details)
- About _select top X by intensity/mz_ on what data do you see this operating (``Spectrum``, ``Scores.scores``, maybe something else)?
- I'll fix the conflict

**Comment by jspaaks** [comment]
The conflict is resolved, but now prospector will complain about it.

File: tests/test_select_by_mz.py (hunk 9 above)
Comment (jspaaks) [reply tests/test_select_by_mz.py]: renamed to ``test_select_by_mz_no_parameters_2`` and moved it adjacent to ``test_select_by_mz_no_parameters_1`` at the top of the file.

File: tests/test_select_by_intensity.py (hunk 3 above)
Comment (jspaaks) [reply tests/test_select_by_intensity.py]: I don't want to do this for unit test functions

File: tests/test_select_by_intensity.py (hunk 4 above)
Comment (jspaaks) [reply tests/test_select_by_intensity.py]: updated to use numpy ndarray.size

File: tests/test_select_by_intensity.py (hunk 6 above)
Comment (jspaaks) [reply tests/test_select_by_intensity.py]: I don't want to do this for unit test functions

File: tests/test_select_by_intensity.py (hunk 8 above)
Comment (jspaaks) [reply tests/test_select_by_intensity.py]: changed it to ``test_select_by_intensity_no_parameters_2``

**Comment by jspaaks** [comment]
OK ladies. Enough chit-chat, I'm merging this.

**Pull request merged**